### PR TITLE
Make ``global-variable-not-assigned`` check local scope

### LINKED ISSUE #330: Two issues with overring functions in the global namespace
Originally reported by: **Antony Lee (BitBucket: [anntzer](http://bitbucket.org/anntzer), GitHub: @anntzer?)**

---

```
#!python
func1 = None
def func2(): pass

def override_func():
    global func1, func2
    def func1(): pass
    func2 = None
```

Pylint gives various spurious warnings here in `override_func`: it does not realize that `def func1():...` assigns a new value at the global level too, and thinks that there is no previous binding for `func2` before overriding it.

---
- Bitbucket: https://bitbucket.org/logilab/pylint/issue/330

### LINKED ISSUE #1375: False negative global-variable-not-assigned (while respective functional test passes correctly)
### Steps to reproduce
Try to lint following code:
```python
a = 1
def foo():
	global a
	print(a)
```

### Current behavior
```
No config file found, using default configuration
************* Module test
W:  3, 0: Found indentation with tabs instead of spaces (mixed-indentation)
W:  4, 0: Found indentation with tabs instead of spaces (mixed-indentation)
C:  1, 0: Missing module docstring (missing-docstring)
C:  1, 0: Invalid constant name "a" (invalid-name)
C:  2, 0: Black listed name "foo" (blacklisted-name)
C:  2, 0: Missing function docstring (missing-docstring)
C:  3, 1: Invalid constant name "a" (invalid-name)
W:  3, 1: Using the global statement (global-statement)
```

### Expected behavior
```
W:  3, 1: Using global for 'a' but no assignment is done (global-variable-not-as
signed)
```

### pylint --version output
```
pylint 1.7.0,
astroid 1.5.0
Python 3.6.0 (v3.6.0:41df79263a11, Dec 23 2016, 07:18:10) [MSC v.1900 32 bit (In
tel)]
```

It is reported correctly in pylint 1.6.5.
```
No config file found, using default configuration
************* Module test
W:  3, 0: Found indentation with tabs instead of spaces (mixed-indentation)
W:  4, 0: Found indentation with tabs instead of spaces (mixed-indentation)
C:  1, 0: Missing module docstring (missing-docstring)
C:  1, 0: Invalid constant name "a" (invalid-name)
C:  2, 0: Black listed name "foo" (blacklisted-name)
C:  2, 0: Missing function docstring (missing-docstring)
W:  3, 1: Using global for 'a' but no assignment is done (global-variable-not-as
signed)
C:  3, 1: Invalid constant name "a" (invalid-name)
```
```
__main__.py 1.6.5,
astroid 1.4.9
Python 3.5.3 (v3.5.3:1880cb95a742, Jan 16 2017, 15:51:26) [MSC v.1900 32 bit (In
tel)]
```

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Add a ChangeLog entry describing what your PR does.
- [x] If it's a new feature, or an important bug fix, add a What's New entry in
      `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description

This checker now checks whether the names after the global keyword are reassigned in the local scope.
I created a small `utils` function that checks whether the name appears in any of the `AssignName` nodes in the scope.

The updated test only throws an error for 2 of the 4 variables. Not quite sure what that is about and we might want to make another issue/PR to fix that. However, since the change did not break any of the tests the file is designed for I guessed it was okay for now. This PR does fix the original issue.

This closes #1375
This closes #330

### CODE REVIEW

File: pylint/checkers/utils.py [hunk 1]
```diff
@@ -1558,3 +1558,11 @@ def is_node_in_guarded_import_block(node: nodes.NodeNG) -> bool:
...
         node.parent.is_sys_guard() or node.parent.is_typing_guard()
     )
+
+
+def is_reassigned_after_current(node: nodes.NodeNG, varname: str) -> bool:
+    """Check if the given variable name is reassigned in the same scope after the current node"""
+    for assign_node in node.scope().nodes_of_class(nodes.AssignName):
+        if assign_node.name == varname and node.lineno < assign_node.lineno:
+            return True
+    return False
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/utils.py:1568]:
```suggestion
    return any(
        a.name == varname and  a.lineno > node.lineno
        for a in node.scope().nodes_of_class(nodes.AssignName)
    )
```

**Review by Pierre-Sassoulas — APPROVED** [review]
I have a small nitpick but this is great :)

**Comment by DanielNoord** [comment]
Sorry for adding changes after your review, but I stumbled upon #330 and it felt weird not to fix it in the same PR.
Basically the latest commit adds additional checks for function redefinitions (both on previous constants and functions) similar to how the first iteration checked for variable assignments.

Because #330 also indicated that function definition in the module scope were not picked up on by the check for `global-variable-undefined` I also added it so we can close that issue.
Thinking about it that might warrant an additional changelog entry.. Let me know if I should add it!

**Comment by Pierre-Sassoulas** [comment]
Don't worry, thank you for fixing another problem :)

**Pull request merged**